spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/CassandraConfSetting.scala

```diff
@@ -0,0 +1,190 @@
+          throw new RuntimeException(s"Missing cluster $c Cassandra connection conf"))
+      case _ => defaultConf
+    }
+  }
+
+  override def addTableConf(
+               table: String,
+               keyspace: String,
+               cluster: Option[String],
+               conf: CassandraConnectorConf): Unit = {
+
+    throw new UnsupportedOperationException("Add table level CassandraConnectorConf is not supported")
+  }
+
+  override def addKeyspaceLevelConf(
+               keyspace: String,
+               cluster: Option[String],
+               conf: CassandraConnectorConf): Unit = {
+
+    throw new UnsupportedOperationException("Add keyspace level CassandraConnectorConf is not supported")
+  }
+
+  override def removeTableLevelConf(
+               table: String,
+               keyspace: String,
+               cluster: Option[String]): Unit = {
+
+    throw new UnsupportedOperationException("remove table level CassandraConnectorConf is not supported")
+  }
+
+  override def removeKeyspaceLevelConf(
+               keyspace: String,
+               cluster: Option[String]) : Unit = {
+
+    throw new UnsupportedOperationException("remove keyspace level CassandraConnectorConf is not supported")
+  }
+
+  override def getConf(
+               table: String,
+               keyspace: String,
+               cluster: Option[String],
+               defaultConf: CassandraConnectorConf): CassandraConnectorConf = {
+
+    throw new UnsupportedOperationException("get table/keyspace level CassandraConnectorConf is not supported")
+
+  }
+}
+
+/**
+ * A local cache to store user defined Cassandra configuration settings. It includes
+ * per cluster level, per keyspace level and per table level settings. It also allows to
+ * update the settings. It's mainly used by Spark SQL, so user can define all kind of
+ * Cassandra configuration settings, and SQL query automatically get the settings to
+ * connect to, read from and write to Cassandra.
+ *
+ * Potentially it could be used by other components as well.
+ *
+ * If it's difficult to isolate the settings for your application, you still could pass
+ * explicitly those settings to the methods.
+ */
+class CassandraConfSetting[T] {
+
+  import CassandraConfSetting._
+
+  /** Low lock contention so use [[java.util.Collections.synchronizedMap]] */
+  protected val clusterConf = java.util.Collections.synchronizedMap(
+    new java.util.HashMap[String, T]())
+  private val keyspaceConf = java.util.Collections.synchronizedMap(
+    new java.util.HashMap[Seq[String], T]())
+  private val tableConf = java.util.Collections.synchronizedMap(
+    new java.util.HashMap[Seq[String], T]())
+
+  /** Add table level configuration settings. Set cluster to None for a single cluster */
+  def addTableConf(
+      table: String,
+      keyspace: String,
+      cluster: Option[String],
+      conf: T): Unit = {
+
+    cluster match {
+      case Some(c) => validateClusterName(c)
```

Comment:
Also, why are you validating the cluster name, but you are *not* validating keyspace and table name?

Side note: This is one of the reasons we should move away from using raw Strings for referencing table/keyspace/cluster names and use value classes with proper validation. Then you'd not be able to construct an invalid table/keyspace/cluster name at all. This is for a separate ticket though.